File: 01_Array_And_Hashing/top_k_frequent_elements.py
class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:
        # Dictionary to store the count of each number
        count = {}

        # List to store numbers grouped by frequency
        freq = [[] for i in range(len(nums) + 1)]

        # Count the occurrences of each number in the input list
        for n in nums:
            count[n] = 1 + count.get(n, 0)

        # Group numbers by frequency in the 'freq' list
        for n, c in count.items():
            freq[c].append(n)

        # Result list to store the top k frequent numbers
        res = []

        # Iterate through the 'freq' list in reverse order
        for i in range(len(freq) - 1, 0, -1):
            # Iterate through the numbers in the current frequency group
            for n in freq[i]:
                # Append the number to the result list
                res.append(n)
                # Check if the result list has reached the desired length k
                if len(res) == k:
                    return res

        # The function will only reach this point if k is greater than the number of unique elements in the input list
        return res
    

# implementation
answer = Solution()
print("Leetcode 347: Top K Frequent Elements\n")

# Example 1
example_1 = [1, 1, 1, 2, 2, 3]
k1 = 2
print(f"Example 1: nums = {example_1}, k = {k1}")
print("Output:", answer.topKFrequent(example_1, k1))

# Example 2
example_2 = [1]
k2 = 1
print(f"Example 2: nums = {example_2}, k = {k2}")
print("Output:", answer.topKFrequent(example_2, k2))

File: 01_Array_And_Hashing/top_k_frequent_elements.go
package main

import (
	"fmt"
)

func topKFrequent(nums []int, k int) []int {
	// Dictionary to store the count of each number
	count := make(map[int]int)

	// List to store numbers grouped by frequency
	freq := make([][]int, len(nums)+1)

	// Count the occurrences of each number in the input list
	for _, num := range nums {
		count[num]++
	}

	// Group numbers by frequency in the 'freq' list
	for index, num := range count {
		freq[num] = append(freq[num], index)
	}

	// Result list to store the top k frequent numbers
	result := make([]int, 0, k)

	// Iterate through the 'freq' list in reverse order
	for index := len(freq) - 1; index > 0; index-- {
		// Iterate through the numbers in the current frequency group
		for _, num := range freq[index] {
			// Append the number to the result list
			result = append(result, num)
			// Check if the result list has reached the desired length k
			if len(result) == k {
				return result
			}
		}
	}

	// The function will only reach this point if k is greater than the number of unique elements in the input list
	return result
}

// implementation

func main() {
	fmt.Println("Leetcode 347: Top K Frequent Elements\n")

	// Example 1
	example1 := []int{1, 1, 1, 2, 2, 3}
	k1 := 2
	fmt.Printf("Example 1: nums = %v, k = %v\n", example1, k1)
	fmt.Println("Output:", topKFrequent(example1, k1))

	// Example 2
	example2 := []int{1}
	k2 := 1
	fmt.Printf("Example 1: nums = %v, k = %v\n", example2, k2)
	fmt.Println("Output:", topKFrequent(example2, k2))

}
